### PR TITLE
Fix #1604 - pass SummaryStyle to ParameterInstance.ToDisplayText()

### LINKED ISSUE #1604: Define custom names for [ParamsSource] classes
# Problem
When using [ParamSource], the class names are often not the names that I want to use for the column entries. For example, I have the following code:
```c#
[ParamsSource(nameof(Serializers))]
public ISerializerTarget Serializer;

public IEnumerable<ISerializerTarget> Serializers => new ISerializerTarget[]
{
	new MessagePackCSharpTarget(),
	new Lz4MessagePackCSharpTarget(),
	new NetSerializerTarget(),
	new MsgPackTarget(),
	new JsonSerializerTarget(),
	new ProtobufNetTarget(),
};
```

Since the classes `MessagePackCSharpTarget` and all others are within the namespace `DotNetSerializationBenchmark` I get the following entries for the benchmark results:
|      Method |       Runtime |           Serializer | ... |
|------------ |-------------- |--------------------- |----------:|
|   **Serialize** | **.NET Core 3.1** | **DotNe(...)arget [49]** | ... |
| Deserialize | .NET Core 3.1 | DotNe(...)arget [49] | ... |
|   Serialize | .NET Core 5.0 | DotNe(...)arget [49] | ... |
| Deserialize | .NET Core 5.0 | DotNe(...)arget [49] | ... |
|   **Serialize** | **.NET Core 3.1** | **DotNe(...)arget [55]** | ... |
| Deserialize | .NET Core 3.1 | DotNe(...)arget [55] | ... |
|   Serialize | .NET Core 5.0 | DotNe(...)arget [55] | ... |
| Deserialize | .NET Core 5.0 | DotNe(...)arget [55] | ... |
|   **Serialize** | **.NET Core 3.1** | **DotNe(...)arget [52]** | ... |
| Deserialize | .NET Core 3.1 | DotNe(...)arget [52] | ... |
| ...           | ...               | ...                      |  ... |

Since everything ends with Target and the namespace is included in the name, it is not readable at all anymore. Note, that `summaryStyle.MaxParameterColumnWidth` is set to `100` and everything is within the same namespace.

## Solution define rules
Use attributes to define how classes / parameters will show up in benchmarks. e.g.
```c#
namespace DotNetSerializationBenchmark
{
	[Display(Name="JsonSerializer")]
	internal class JsonSerializerTarget : ASerializerTarget<MemoryStream>
	{
```
I think, this approach would also offer additional flexibility to customize RPlots (#1558).

## Solution change defaults
* `summaryStyle.MaxParameterColumnWidth` has an effect on the string creation for ParamSource (Is it intended to not be that way or is that a bug?)
* Don't include namespace if the class is in the same namespace as the benchmark

### PR DESCRIPTION
**Pull request by JohannesDeml**
Fix proposal for #1604 

By enabling to pass the summary style to ParameterInstances, it is possible to have different max column widths for different exporters. It also fixes the problem when not parsing the config through the run method (at which point the constructor for ParameterInstance is called).

### CODE REVIEW

**Comment by dnfadmin** [comment]
[CLA assistant check](https://cla.dotnetfoundation.org/dotnet/BenchmarkDotNet?pullRequest=1609) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/dotnet/BenchmarkDotNet?pullRequest=1609) before we can accept your contribution.<br/><br/>:x: JohannesDeml [sign now](https://cla.dotnetfoundation.org/dotnet/BenchmarkDotNet?pullRequest=1609)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/dotnet/BenchmarkDotNet?pullRequest=1609) it.</sub>

File: src/BenchmarkDotNet/Parameters/ParameterInstance.cs [hunk 1]
```diff
@@ -55,6 +55,10 @@ public string ToDisplayText(CultureInfo cultureInfo)
             }
         }
 
+        public string ToDisplayText(SummaryStyle summary) => ToDisplayText(summary.CultureInfo, summary.MaxParameterColumnWidth);
```
Comment (adamsitnik) [inline src/BenchmarkDotNet/Parameters/ParameterInstance.cs:58]: From the CI logs, it looks that `summary` can be `null` here? Could you please fix it (by checking if we can avoid passing `null` here or making sure the method works fine when a `null` is passed)?

```log
Error Message:
 System.NullReferenceException : Object reference not set to an instance of an object.
Stack Trace:
   at BenchmarkDotNet.Parameters.ParameterInstance.ToDisplayText(SummaryStyle summary) in C:\projects\benchmarkdotnet\src\BenchmarkDotNet\Parameters\ParameterInstance.cs:line 58
   at BenchmarkDotNet.Columns.ParamColumn.GetValue(Summary summary, BenchmarkCase benchmarkCase) in C:\projects\benchmarkdotnet\src\BenchmarkDotNet\Columns\ParamColumn.cs:line 20
   at System.Linq.Enumerable.SelectArrayIterator`2.ToArray()
   at System.Linq.Enumerable.ToArray[TSource](IEnumerable`1 source)
   at System.Linq.Enumerable.SelectArrayIterator`2.ToArray()
   at System.Linq.Enumerable.ToArray[TSource](IEnumerable`1 source)
   at BenchmarkDotNet.Reports.SummaryTable..ctor(Summary summary, SummaryStyle style) in C:\projects\benchmarkdotnet\src\BenchmarkDotNet\Reports\SummaryTable.cs:line 60
   at BenchmarkDotNet.Reports.Summary..ctor(String title, ImmutableArray`1 reports, HostEnvironmentInfo hostEnvironmentInfo, String resultsDirectoryPath, String logFilePath, TimeSpan totalTime, CultureInfo cultureInfo, ImmutableArray`1 validationErrors) in C:\projects\benchmarkdotnet\src\BenchmarkDotNet\Reports\Summary.cs:line 60
   at BenchmarkDotNet.Running.BenchmarkRunnerClean.Run(BenchmarkRunInfo benchmarkRunInfo, Dictionary`2 buildResults, IResolver resolver, ILogger logger, List`1 artifactsToCleanup, String resultsFolderPath, String logFilePath, StartedClock& runChronometer) in C:\projects\benchmarkdotnet\src\BenchmarkDotNet\Running\BenchmarkRunnerClean.cs:line 134
   at BenchmarkDotNet.Running.BenchmarkRunnerClean.Run(BenchmarkRunInfo[] benchmarkRunInfos) in C:\projects\benchmarkdotnet\src\BenchmarkDotNet\Running\BenchmarkRunnerClean.cs:line 76
   at BenchmarkDotNet.Running.BenchmarkRunner.RunWithDirtyAssemblyResolveHelper(Type type, IConfig config) in C:\projects\benchmarkdotnet\src\BenchmarkDotNet\Running\BenchmarkRunnerDirty.cs:line 77
   at BenchmarkDotNet.Running.BenchmarkRunner.RunWithExceptionHandling(Func`1 run) in C:\projects\benchmarkdotnet\src\BenchmarkDotNet\Running\BenchmarkRunnerDirty.cs:line 107
   at BenchmarkDotNet.Running.BenchmarkRunner.Run(Type type, IConfig config) in C:\projects\benchmarkdotnet\src\BenchmarkDotNet\Running\BenchmarkRunnerDirty.cs:line 30
   at BenchmarkDotNet.IntegrationTests.BenchmarkTestExecutor.CanExecute(Type type, IConfig config, Boolean fullValidation) in C:\projects\benchmarkdotnet\tests\BenchmarkDotNet.IntegrationTests\BenchmarkTestExecutor.cs:line 63
   at BenchmarkDotNet.IntegrationTests.ParamsTestStaticPropertyTest.Test() in C:\projects\benchmarkdotnet\tests\BenchmarkDotNet.IntegrationTests\ParamsAttributeStaticTest.cs:line 18
```

File: src/BenchmarkDotNet/Parameters/ParameterInstance.cs [hunk 2]
```diff
@@ -37,7 +37,7 @@ public string ToSourceCode()
                 ? parameter.ToSourceCode()
                 : SourceCodeHelper.ToSourceCode(value);
 
-        public string ToDisplayText(CultureInfo cultureInfo)
+        public string ToDisplayText(CultureInfo cultureInfo, int maxParameterColumnWidth)
```
[the remark this answers is not in the document]
Comment (JohannesDeml) [reply src/BenchmarkDotNet/Parameters/ParameterInstance.cs:40]: Yes, good idea!

File: src/BenchmarkDotNet/Parameters/ParameterInstance.cs [hunk 3]
```diff
@@ -15,13 +15,13 @@ public class ParameterInstance : IDisposable
         [PublicAPI] public ParameterDefinition Definition { get; }
 
         private readonly object value;
-        private readonly int maxParameterColumnWidth;
+        private readonly int defaultMaxParameterColumnWidth;
```
[the remark this answers is not in the document]
Comment (JohannesDeml) [reply src/BenchmarkDotNet/Parameters/ParameterInstance.cs]: True, I didnt think about the existance of `SummaryStyle.DefaultMaxParameterColumnWidth`. Good suggestion, thanks!

File: src/BenchmarkDotNet/Parameters/ParameterInstance.cs (hunk 1 above)
Comment (JohannesDeml) [reply src/BenchmarkDotNet/Parameters/ParameterInstance.cs]: I added a null check, since we do have a valid fallback (by using `maxParameterColumnWidthFromConfig`). This way, the class acts exactly the same way as before, when the SummaryStyle is null, so it won't break anything for users that rely on a null SummaryStyle.
If usage without a summary style should be prohibited, I guess we should change all tests that don't have a Summary style (not that many), to use one and add another test with a null summary style that should throw an error to document that requirement, right?

**Comment by JohannesDeml** [comment]
@adamsitnik Thanks a lot for your guidance! Tests are passing now. If you're okay with the null check, I guess the PR is good to go :)

File: src/BenchmarkDotNet/Parameters/ParameterInstance.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (adamsitnik) [reply src/BenchmarkDotNet/Parameters/ParameterInstance.cs:40]: @JohannesDeml it's still `public`, have you forgot to change it?

File: src/BenchmarkDotNet/Parameters/ParameterInstance.cs [hunk 4]
```diff
@@ -55,8 +55,15 @@ public string ToDisplayText(CultureInfo cultureInfo)
...
 
+        public string ToDisplayText(SummaryStyle summary)
+        {
+            return summary != null ? ToDisplayText(summary.CultureInfo, summary.MaxParameterColumnWidth) : ToDisplayText();
+        }
+
         public string ToDisplayText() => ToDisplayText(CultureInfo.CurrentCulture);
 
+        private string ToDisplayText(CultureInfo cultureInfo) => ToDisplayText(cultureInfo, maxParameterColumnWidthFromConfig);
+
```
Comment (adamsitnik) [inline src/BenchmarkDotNet/Parameters/ParameterInstance.cs:66]: is there only one usage of `ToDisplayText(CultureInfo cultureInfo)`? If so, can we inline it?


```suggestion

        public string ToDisplayText() => ToDisplayText(cultureInfo, maxParameterColumnWidthFromConfig);
```

File: src/BenchmarkDotNet/Parameters/ParameterInstance.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (JohannesDeml) [reply src/BenchmarkDotNet/Parameters/ParameterInstance.cs]: Oh, sorry! I changed `ToDisplayText(CultureInfo cultureInfo)` before (since that method is also not used outside of the class). Fixed access for `ToDisplayText(CultureInfo cultureInfo, int maxParameterColumnWidth)` now.

File: src/BenchmarkDotNet/Parameters/ParameterInstance.cs (hunk 4 above)
Comment (JohannesDeml) [reply src/BenchmarkDotNet/Parameters/ParameterInstance.cs]: 👍

File: src/BenchmarkDotNet/Parameters/ParameterInstance.cs [hunk 5]
```diff
@@ -55,8 +55,12 @@ public string ToDisplayText(CultureInfo cultureInfo)
             }
         }
 
-        public string ToDisplayText() => ToDisplayText(CultureInfo.CurrentCulture);
+        public string ToDisplayText(SummaryStyle summary)
+        {
+            return summary != null ? ToDisplayText(summary.CultureInfo, summary.MaxParameterColumnWidth) : ToDisplayText();
+        }
 
+        public string ToDisplayText() => ToDisplayText(cultureInfo, maxParameterColumnWidthFromConfig);
```
Comment (adamsitnik) [inline src/BenchmarkDotNet/Parameters/ParameterInstance.cs:63]: I am sorry, my previous suggestion was incomplete

```suggestion
        public string ToDisplayText() => ToDisplayText(CultureInfo.CurrentCulture, maxParameterColumnWidthFromConfig);

```

Comment (JohannesDeml) [reply src/BenchmarkDotNet/Parameters/ParameterInstance.cs]: Sorry, I didn't notice. Thanks for fixing it!

**Review by adamsitnik — APPROVED** [review]
LGTM, thank you very much @JohannesDeml !

**Comment by JohannesDeml** [comment]
Awesome, thank you for guiding me through me my first PR to this repo!
